### ISSUE TITLE
Drop in accuracy after using merge_bn

### ISSUE THREAD

**Issue by MohamedA95** [Xilinx/brevitas#415]
Hello everyone,
I am training a Lenet model built using brevitas. After training, I use `merge_bn` to merge BatchNorm layers. But I notice a 5~7% drop in test accuracy between the merged & unmerged models. While I do not get any drop if I build the model using torch layers. I tried to fuse the model using `brevitas.nn.utils.merge_bn` & `torch.nn.utils.fusion.fuse_conv_bn_eval` but got the same result. Any idea how can I solve this? I tried with both dev & master branches & got the same result. A similar drop is present in other networks like VGG & U-net.
This is how I define the model 
```python
WEIGHT_QUANTIZER = Int8WeightPerTensorFixedPoint
INPUT_QUANTIZER = Int8ActPerTensorFixedPoint
ACT_QUANTIZER = None
BIAS_QUANTIZER = Int8Bias
ENABLE_BIAS_QUANT = True

class QuantLeNet(Module):
    def __init__(self, bit_width=8, batchnorm=True, enable_bias_quant=ENABLE_BIAS_QUANT,num_classes=10):
        super(QuantLeNet, self).__init__()
        self.features = nn.Sequential(
            *make_quant_conv2d(3, 32, 3, bit_width, bias=True, return_quant_tensor=not batchnorm,
                               batchnorm=batchnorm, enable_bias_quant=enable_bias_quant),
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2,stride=2),
            *make_quant_conv2d(32, 64, 3, bit_width, return_quant_tensor=not batchnorm,
                               batchnorm=batchnorm, enable_bias_quant=enable_bias_quant),
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2,stride=2)
        )
        self.classifier = nn.Sequential(
            make_quant_linear(in_features=64*6*6, out_features=120, bit_width=bit_width, bias=True, enable_bias_quant=enable_bias_quant),
            nn.ReLU(),
            make_quant_linear(in_features=120, out_features=84, bit_width=bit_width, bias=True, enable_bias_quant=enable_bias_quant,input_quant=None),
            nn.ReLU(),
            make_quant_linear(in_features=84, out_features=num_classes, bit_width=bit_width, bias=True, enable_bias_quant=enable_bias_quant,input_quant=None))

    def forward(self, x):
        out = self.features(x)
        out = out.reshape(out.shape[0], -1)
        out = self.classifier(out)
        return out


def make_quant_conv2d(in_channels,
                      out_channels,
                      kernel_size,
                      bit_width,
                      stride=1,
                      padding=0,
                      dilation=1,
                      groups=1,
                      bias=True,
                      batchnorm=False,
                      weight_quant=WEIGHT_QUANTIZER,
                      bias_quant=BIAS_QUANTIZER,
                      input_quant=INPUT_QUANTIZER,
                      output_quant=INPUT_QUANTIZER,
                      return_quant_tensor=True,
                      enable_bias_quant=ENABLE_BIAS_QUANT,
                      weight_scaling_per_output_channel=False):
    layers=[]
    layers.append(qnn.QuantConv2d(in_channels,
                              out_channels,
                              kernel_size=kernel_size,
                              stride=stride,
                              padding=padding,
                              dilation=dilation,
                              groups=groups,
                              bias=bias,
                              weight_quant=weight_quant,
                              bias_quant=bias_quant,
                              input_quant=input_quant,
                              output_quant=output_quant,
                              weight_bit_width=bit_width,
                              input_bit_width=bit_width,
                              return_quant_tensor=return_quant_tensor,
                              compute_output_bit_width=bias and enable_bias_quant,
                              compute_output_scale=bias and enable_bias_quant,
                              weight_scaling_per_output_channel=weight_scaling_per_output_channel))
    layers[0].cache_inference_quant_bias = True
    if(batchnorm):
        layers.append(nn.BatchNorm2d(out_channels))
    return layers


def make_quant_linear(in_features,
                      out_features,
                      bit_width,
                      bias,
                      weight_quant=WEIGHT_QUANTIZER,
                      input_quant=INPUT_QUANTIZER,
                      bias_quant=BIAS_QUANTIZER,
                      return_quant_tensor=True,
                      output_quant=INPUT_QUANTIZER,
                      enable_bias_quant=ENABLE_BIAS_QUANT):
    quant_liner = qnn.QuantLinear(in_features, out_features,
                           weight_quant=weight_quant,
                           bias_quant=bias_quant,
                           input_quant=input_quant,
                           input_bit_width=bit_width,
                           return_quant_tensor=return_quant_tensor,
                           output_quant=output_quant,
                           output_bit_width=bit_width,
                           bias=bias,
                           compute_output_bit_width=bias and enable_bias_quant,
                           compute_output_scale=bias and enable_bias_quant,
                           weight_bit_width=bit_width)
    quant_liner.cache_inference_quant_bias = True
    return quant_liner
```

And this is how I fuse the model:

```python
            if isinstance(i, qnn.QuantConv2d) or isinstance(i, nn.Conv2d):
                bn = next(features_iter)
                assert isinstance(bn, nn.BatchNorm2d), "The layer after QuantConv2d is not nn.BatchNorm2d"
                merge_bn(i, bn)
                fused_model.features[fusedindex].load_state_dict(i.state_dict())
```

**Comment by volcacius** [comment]
Hi Mohamed,

There was a bug in merging with an existing bias, so that might one source of problem, it's fixed in dev since last week. You also have per-channel scaling disabled, if you merge a BN into that you will notice a non trivial numerical difference. Another source of difference is the fact that you are using power of two scale factors (fixed point quantizers) and not floating point ones.  
Enabling per channel scaling should make the biggest difference. Any other source of error, you can always retrain after merging to regain the lost accuracy. Keep in mind that in general for quantized neural networks batch norm merging is a lossy transformation, except for per-channel floating point scale factors.

**Comment by MohamedA95** [comment]
Hi Alessandro,
Thanks for your reply, I deeply appreciate it. Just a little bit of
context, my final goal is to implement the model in HLS using finn-hlslib.
As far as I understand enabling per channel scaling would give each channel
a different it's own scale hence it's own bit width. Which would be tricky
to implement in HLS, please correct me if I'm wrong. I am using fixed point
quantizers because I want to implement the model on HW with specific bit
width. I will try retraining. Thanks again for your help.

**Comment by volcacius** [comment]
Scale and bit width are two different things. FINN supports layer-wise bit width and channel-wise floating point scale, as well as floating point biases and unmerged batch norm. The FINN compiler handles all the merging and transformation logic to map that to optimized layers in the finn-hlslib. I would go through the compiler then, rather than directly use the finn-hlslib, otherwise you are going to have to handle all the mapping and transformation by hand.

**Comment by MohamedA95** [comment]
Hi Alessandro,
Sorry, that's my mistake. I mean that under fixed-point a different [scale](https://github.com/Xilinx/brevitas/issues/271#issuecomment-800987165) would mean a different fraction part width. Overall I get your point thanks a lot for your support.
